File: android/app/src/main/java/me/masonasons/fastsm/push/PushManager.kt
package me.masonasons.fastsm.push

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import androidx.core.content.ContextCompat
import com.google.firebase.FirebaseApp
import com.google.firebase.messaging.FirebaseMessaging
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

/**
 * The Android side of push notifications: the Firebase registration token, the
 * relay registration that turns it into a Web Push endpoint, and the on/off
 * preference. This is the OS-level plumbing an Android app must do itself.
 *
 * Everything portable stays in the core: it owns the Mastodon
 * /api/v1/push/subscription call for every signed-in account (push_subscribe /
 * push_unsubscribe), exactly as on iOS. The app only supplies the endpoint and
 * the public key material.
 *
 * Delivery path: Mastodon posts an encrypted Web Push body to the relay, the
 * relay forwards it to Firebase Cloud Messaging as a data-only message, and
 * [FastSmMessagingService] decrypts it on-device. The relay never sees the
 * contents.
 */
object PushManager {

    private const val TAG = "FastSmPush"

    private const val PREFS = "push"
    private const val ENABLED = "enabled"
    private const val ENDPOINT_ID = "endpoint_id"

    /** The self-hosted relay that bridges Mastodon Web Push to FCM and APNs. */
    private const val RELAY_BASE = "https://push.brynify.me"

    /**
     * Shared secret the relay's /register requires. Low-secrecy: it only gates
     * casual abuse of the registration endpoint, and mirrors
     * RELAY_REGISTER_TOKEN in the relay's config.
     */
    private const val RELAY_REGISTER_TOKEN =
        "be354d899a9bf6af74578ed7a631f2a916e25348d1e07f5fb4153121a13d90be"

    private val http = OkHttpClient()

    /** What the core needs to subscribe an account with its Mastodon server. */
    data class Subscription(val endpoint: String, val p256dh: String, val auth: String)

    /** Whether the user has turned push on (persisted). */
    fun isEnabled(context: Context): Boolean = prefs(context).getBoolean(ENABLED, false)

    fun setEnabled(context: Context, on: Boolean) {
        prefs(context).edit().putBoolean(ENABLED, on).apply()
    }

    /**
     * Whether this build can do push at all. Firebase needs a google-services.json
     * at build time; a build without one (a fork, or a local build without the
     * config) still runs, it just has no push.
     */
    fun isAvailable(context: Context): Boolean =
        runCatching { FirebaseApp.getApps(context).isNotEmpty() }.getOrDefault(false)

    /** Whether Android will let us post notifications (API 33+ asks the user). */
    fun hasNotificationPermission(context: Context): Boolean =
        Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU ||
            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) ==
            PackageManager.PERMISSION_GRANTED

    /** The runtime permission to ask for on API 33+, or null when none is needed. */
    fun notificationPermission(): String? =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            Manifest.permission.POST_NOTIFICATIONS
        } else {
            null
        }

    /**
     * Turn push on: fetch the Firebase token, register it with the relay for a
     * per-device endpoint, and hand that endpoint plus our Web Push public keys
     * back so the caller can pass them to the core.
     *
     * [onResult] gets null if anything failed; it runs on a background thread.
     */
    fun enable(context: Context, onResult: (Subscription?) -> Unit) {
        setEnabled(context, true)
        register(context, onResult)
    }

    /**
     * Turn push off. The caller sends push_unsubscribe to the core (which drops
     * the Mastodon subscriptions); here we only stop this device receiving.
     */
    fun disable(context: Context) {
        setEnabled(context, false)
        // Drop the Firebase token so nothing more is delivered even if a
        // subscription lingers server-side.
        runCatching { FirebaseMessaging.getInstance().deleteToken() }
            .onFailure { Log.w(TAG, "deleteToken failed", it) }
    }

    /**
     * Called at app start: if push was left on, re-register and re-subscribe.
     * Mastodon subscriptions can lapse and the POST simply replaces, so this is
     * safe to repeat.
     */
    fun refreshIfEnabled(context: Context, onResult: (Subscription?) -> Unit) {
        if (isEnabled(context)) register(context, onResult)
    }

    /**
     * Re-point the existing endpoint at a new Firebase token, without touching
     * the Mastodon subscription. Firebase can rotate a token at any time; because
     * the endpoint id is stable, updating the relay's mapping is enough and the
     * subscription Mastodon already holds keeps delivering. Used from
     * [FastSmMessagingService.onNewToken], where the core isn't running.
     */
    fun onTokenRotated(context: Context, token: String) {
        if (!isEnabled(context)) return
        registerWithRelay(context, token) { endpoint ->
            if (endpoint == null) Log.w(TAG, "relay re-registration after token rotation failed")
        }
    }

    // ---- internals ----

    private fun register(context: Context, onResult: (Subscription?) -> Unit) {
        if (!isAvailable(context)) {
            Log.w(TAG, "no Firebase configuration in this build; push unavailable")
            onResult(null)
            return
        }
        val messaging = runCatching { FirebaseMessaging.getInstance() }.getOrNull()
        if (messaging == null) {
            onResult(null)
            return
        }
        messaging.token.addOnCompleteListener { task ->
            val token = task.result
            if (!task.isSuccessful || token.isNullOrEmpty()) {
                Log.w(TAG, "no Firebase token", task.exception)
                onResult(null)
                return@addOnCompleteListener
            }
            registerWithRelay(context, token) { endpoint ->
                if (endpoint == null) {
                    onResult(null)
                } else {
                    val keys = WebPushKeys.ensure(context)
                    onResult(Subscription(endpoint, keys.p256dh, keys.auth))
                }
            }
        }
    }

    /**
     * Trade the Firebase token for the relay endpoint URL Mastodon will post to.
     * The endpoint id is remembered and sent back on later registrations, so a
     * device keeps one endpoint instead of stranding a new one each time.
     */
    private fun registerWithRelay(
        context: Context,
        token: String,
        onEndpoint: (String?) -> Unit,
    ) {
        val body = JSONObject()
            .put("device_token", token)
            .put("platform", "fcm")
        prefs(context).getString(ENDPOINT_ID, null)?.let { body.put("endpoint_id", it) }

        val request = Request.Builder()
            .url("$RELAY_BASE/register")
            .header("Authorization", "Bearer $RELAY_REGISTER_TOKEN")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()

        http.newCall(request).enqueue(object : okhttp3.Callback {
            override fun onFailure(call: okhttp3.Call, e: java.io.IOException) {
                Log.w(TAG, "relay register failed", e)
                onEndpoint(null)
            }

            override fun onResponse(call: okhttp3.Call, response: okhttp3.Response) {
                response.use {
                    val text = it.body?.string().orEmpty()
                    if (!it.isSuccessful) {
                        Log.w(TAG, "relay register: HTTP " + it.code)
                        onEndpoint(null)
                        return
                    }
                    val endpoint = runCatching { JSONObject(text).optString("endpoint") }
                        .getOrDefault("")
                    if (endpoint.isEmpty()) {
                        Log.w(TAG, "relay register: no endpoint in response")
                        onEndpoint(null)
                        return
                    }
                    prefs(context).edit()
                        .putString(ENDPOINT_ID, endpoint.substringAfterLast("/"))
                        .apply()
                    onEndpoint(endpoint)
                }
            }
        })
    }

    private fun prefs(context: Context) =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
}

File: ios/src/PushManager.swift
//
//  PushManager.swift
//
//  The iOS side of push notifications: asking the user's permission,
//  registering with APNs, and holding the resulting device token. This is the
//  OS-level plumbing a Swift app must do itself. Everything portable — the Web
//  Push keypair, the Mastodon /api/v1/push/subscription call, and the strings a
//  notification shows — lives in the core so every front end shares it.
//
//  P1 (this file) only registers and captures the token; it is deliberately NOT
//  invoked at launch, since prompting for notifications before the relay exists
//  would lead nowhere. A Notifications setting will drive it once the backend is
//  live (P5). `onDeviceToken` is where the core subscription hooks in (P2).
//

import Foundation
import UIKit
import UserNotifications
import os

final class PushManager: NSObject {
    static let shared = PushManager()

    private let log = Logger(subsystem: "me.masonasons.FastSMRW", category: "push")

    /// The most recent APNs device token (lowercase hex), once registration has
    /// succeeded. Handed to the core Web Push subscription at P2.
    private(set) var deviceTokenHex: String?

    /// Fired when a fresh device token arrives, after the relay+Mastodon
    /// subscription is kicked off. Optional hook for UI.
    var onDeviceToken: ((String) -> Void)?

    // The self-hosted relay that bridges Mastodon Web Push to APNs.
    private static let relayBase = "https://push.brynify.me"
    // Shared secret the relay's /register requires (low-secrecy; gates casual
    // abuse). Mirrors RELAY_REGISTER_TOKEN in the relay's config.
    private static let relayRegisterToken =
        "be354d899a9bf6af74578ed7a631f2a916e25348d1e07f5fb4153121a13d90be"

    private static let enabledKey = "push_enabled"

    /// Whether the user has turned push on (persisted).
    var isEnabled: Bool { UserDefaults.standard.bool(forKey: Self.enabledKey) }

    private weak var state: AppState?

    /// Set while an explicit enable is in flight, so the core speaks the result
    /// only for a change the user asked for -- not for the renewal at launch.
    private var announceNextResult = false
	private var accountKeys: Set<String> = []

	/// Renew after accounts have loaded, and whenever another account is added.
	@MainActor
	func accountsChanged(state: AppState) {
		let keys = Set(state.accounts.filter { $0.platform == "mastodon" }.map { $0.key })
		let added = !keys.subtracting(accountKeys).isEmpty
		accountKeys = keys
		if added { refreshIfEnabled(state: state) }
	}

    /// Turn push on: remember the choice, ask permission, register with APNs,
    /// and (once the token arrives) subscribe with the relay + Mastodon.
    @MainActor
    func enable(state: AppState) {
        self.state = state
        UserDefaults.standard.set(true, forKey: Self.enabledKey)
        announceNextResult = true
        requestAuthorizationAndRegister()
    }

    /// Turn push off: drop the Mastodon subscription and stop APNs.
    @MainActor
    func disable(state: AppState) {
        self.state = state
        UserDefaults.standard.set(false, forKey: Self.enabledKey)
		announceNextResult = false
        state.pushUnsubscribe(announce: true)
        UIApplication.shared.unregisterForRemoteNotifications()
    }

    /// Called at launch: if push was left on, re-register and re-subscribe
    /// (Mastodon subscriptions can lapse; the POST simply replaces).
    @MainActor
    func refreshIfEnabled(state: AppState) {
        guard isEnabled else { return }
        self.state = state
        requestAuthorizationAndRegister()
    }

    /// Ask permission and, if granted, register with APNs. Safe to call more
    /// than once — iOS shows the system prompt only the first time.
    func requestAuthorizationAndRegister() {
        UNUserNotificationCenter.current()
            .requestAuthorization(options: [.alert, .sound, .badge]) { [weak self] granted, error in
                if let error {
                    self?.log.error("push auth error: \(error.localizedDescription, privacy: .public)")
                }
                guard granted else {
                    self?.log.info("push authorization denied")
                    return
                }
                Task { @MainActor in UIApplication.shared.registerForRemoteNotifications() }
            }
    }

    // MARK: Called from AppDelegate's OS callbacks (on the main thread).

    func didRegister(deviceToken: Data) {
        let hex = deviceToken.map { String(format: "%02x", $0) }.joined()
        deviceTokenHex = hex
        log.info("APNs device token received")
        if isEnabled { subscribe(deviceTokenHex: hex) }
        onDeviceToken?(hex)
    }

    func didFailToRegister(error: Error) {
        log.error("APNs registration failed: \(error.localizedDescription, privacy: .public)")
    }

    // MARK: - relay registration + Mastodon subscription

    /// Register the device token with the relay to get a per-device endpoint,
    /// then hand that endpoint + our Web Push public keys to the core, which
    /// subscribes the selected Mastodon account.
    private func subscribe(deviceTokenHex: String) {
        let keys = WebPushKeys.ensure()
        var req = URLRequest(url: URL(string: Self.relayBase + "/register")!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer " + Self.relayRegisterToken, forHTTPHeaderField: "Authorization")
        let body: [String: String] = ["device_token": deviceTokenHex,
                                       "environment": Self.apnsEnvironment()]
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)
        URLSession.shared.dataTask(with: req) { [weak self] data, resp, err in
            guard let self else { return }
            if let err {
                self.log.error("relay register failed: \(err.localizedDescription, privacy: .public)")
                return
            }
            guard let data,
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let endpoint = obj["endpoint"] as? String else {
                let code = (resp as? HTTPURLResponse)?.statusCode ?? -1
                self.log.error("relay register: bad response (\(code, privacy: .public))")
                return
            }
            Task { @MainActor in
				guard self.isEnabled else { return }
				let announce = self.announceNextResult
				self.announceNextResult = false
                self.state?.pushSubscribe(endpoint: endpoint, p256dh: keys.p256dh,
                                          auth: keys.auth, announce: announce)
            }
        }.resume()
    }

    /// Which APNs environment this build's device token belongs to, read from the
    /// embedded provisioning profile. App Store builds carry no profile ->
    /// production.
    static func apnsEnvironment() -> String {
        guard let url = Bundle.main.url(forResource: "embedded", withExtension: "mobileprovision"),
              let data = try? Data(contentsOf: url),
              let text = String(data: data, encoding: .isoLatin1),
              let k = text.range(of: "aps-environment"),
              let s = text.range(of: "<string>", range: k.upperBound..<text.endIndex),
              let e = text.range(of: "</string>", range: s.upperBound..<text.endIndex)
        else { return "production" }
        return text[s.upperBound..<e.lowerBound] == "development" ? "sandbox" : "production"
    }
}

extension PushManager: UNUserNotificationCenterDelegate {
    /// Show pushes even while the app is in the foreground.
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                willPresent notification: UNNotification,
                                withCompletionHandler completionHandler:
                                    @escaping (UNNotificationPresentationOptions) -> Void) {
        completionHandler([.banner, .sound, .list])
    }
}
